python/cudf/cudf/core/tools/numeric.py

```diff
@@ -0,0 +1,236 @@
+            "signed": list(np.typecodes["Integer"]),
+            "unsigned": list(np.typecodes["UnsignedInteger"]),
+        }
+        float_types = list(np.typecodes["Float"])
+        idx = float_types.index(np.dtype(np.float32).char)
+        downcast_type_map["float"] = float_types[idx:]
+
+        type_set = downcast_type_map[downcast]
+
+        for t in type_set:
+            downcast_dtype = np.dtype(t)
+            if downcast_dtype.itemsize <= col.dtype.itemsize:
+                if col.can_cast_safely(downcast_dtype):
+                    col = libcudf.unary.cast(col, downcast_dtype)
+                    break
+
+    if isinstance(arg, (cudf.Series, pd.Series)):
+        return cudf.Series(col)
+    else:
+        col = col.fillna(col.default_na_value())
+        return col.values
+
+
+def _convert_str_col(col, errors, _downcast=None):
+    """
+    Converts a string column to numeric column
+
+    Converts to integer column if all strings are integer-like (isinteger.all)
+    Otherwise, converts to float column if all strings are float-like (
+    isfloat.all)
+
+    If error == 'coerce', fill non-numerics strings with null
+
+    Looks ahead to ``downcast`` parameter, if the float may be casted to
+    integer, then only process in float32 pipeline.
+
+    Parameters
+    ----------
+    col : The string column to convert, must be string dtype
+    errors : {'raise', 'ignore', 'coerce'}, same as ``to_numeric``
+    _downcast : Same as ``to_numeric``, see description for use
+
+    Returns
+    -------
+    Converted numeric column
+    """
+    if not is_string_dtype(col):
+        raise TypeError("col must be string dtype.")
+
+    is_integer = col.str().isinteger()
+    if is_integer.all():
+        return col.as_numerical_column(dtype=np.dtype("i8"))
+
+    # Account for `inf` strings
+    col = _proc_inf_strings(col)
+
+    is_float = col.str().isfloat()
+    if is_float.all():
+        if _downcast in {"unsigned", "signed", "integer"}:
+            warnings.warn(
+                UserWarning(
+                    "Downcasting from float to int will be "
+                    "limited by float32 precision."
+                )
+            )
+            return col.as_numerical_column(dtype=np.dtype("f"))
+        else:
+            return col.as_numerical_column(dtype=np.dtype("d"))
+    else:
+        if errors == "coerce":
+            col = libcudf.string_casting.stod(col)
+            non_numerics = is_integer.binary_operator(
+                "or", is_float
+            ).unary_operator("not")
+            col[non_numerics] = None
+            return col
+        else:
+            raise ValueError("Unable to convert some strings to numerics.")
+
+
+def _proc_inf_strings(col):
+    col = col.str().lower()
+    col = col.str().replace(
+        ["+infinity", "-infinity", "infinity", "+inf", "-inf", "inf"],
+        ["Inf", "-Inf", "Inf", "Inf", "-Inf", "Inf"],
+        regex=False,
+    )
+    return col
```

Comment:
Need to refactor this part due to the undefined order of replacement.

Reply:
Can you provide an example where this would fail?

Reply:
Offline chat with @davidwendt summary: 
- Mapping `["+", "inf", "inity"],["", "Inf", ""]` seems to be the simplest, and order-agnostic.
- Tricks to handle empty strings: use `Series.where` since `replace` doesn't take empty strings argument.
- Could make use of `replace_empty` in the future